tests/test_descriptors.py

```diff
@@ -1,34 +1,163 @@
 import pytest
 
-from bionic.descriptors import DescriptorNode, EntityNode
+from bionic.exception import MalformedDescriptorError
+from bionic.descriptors.parsing import (
+    dnode_from_descriptor,
+    entity_dnode_from_descriptor,
+)
+from bionic.descriptors.ast import EntityNode, TupleNode
+
+from .helpers import assert_re_matches
+
+E = EntityNode
+T = lambda *children: TupleNode(children)  # noqa: E731
+parse = dnode_from_descriptor
+
+
+def check_roundtrip(descs, dnode):
+    """
+    Checks that a group of descriptor strings all parse to the provided descriptor node,
+    and that that dnode can be converted back to the first descriptor string.
+    """
+
+    for desc in descs:
+        assert parse(desc) == dnode
+    assert dnode.to_descriptor() == descs[0]
+
+
+def test_parsing_and_unparsing():
+    check_roundtrip(
+        descs=["x", "(x)", "((x))", " x "], dnode=E("x"),
+    )
+    check_roundtrip(
+        descs=["()", "(())", "( )"], dnode=T(),
```

Comment:
Somewhat related to the single element comment, why should Bionic support empty tuples? I would argue that we disallow it in our grammar instead of later throwing an error about an empty tuple. Unless there is a use case for supporting empty tuples that I might have missed.

Reply:
I do think we should allow empty tuples. The full reasoning is easier to convey in a conversation rather than a comment, but just a few points:

1. There's no reason we would "later throw an error". Everything we do with tuples has intuitive, well-defined behavior for empty tuples as well. (E.g., `flow.get('()')` will obviously just return an empty Python tuple.)

2. It's hard to imagine a real-world situation where you'd want to explicitly construct an empty tuple descriptor. But the same is true for Python tuples, and they're supported. (In fact, a lot of stuff in Python would break if empty tuples weren't allowed.) This is not a complete argument, of course, but I think a lot of the fundamental reasoning carries over. WDYT?

Also, all this applies to singleton tuples as well.